### ISSUE TITLE
Non 1D object can pollute vector data

### ISSUE THREAD

**Issue by sauvikbiswas** [sauvikbiswas/vectorlib#5]
If a NumPy object (or any structure with mappable float values) of an arbitrary dimension is fed into the constructor, the initialisation occurs as is, with a dataset that cannot use most of the functions.

``` python
# Constructor & Destructor
def __init__(self, listarr):
    try:
        self.data = map(float, listarr)
        return
    except ValueError:
        raise ValueError('Cannot obtain a float representation')
```

No checks on `listarr` is made to ensure that it is 1D. One possibility is to linearize the structure before float mapping.

**Comment by sauvikbiswas** [comment]
One way is to convert it and check if all entries are float. That way we can avoid linearisation. In fact, linearisation might not be necessary. The code should raise a `ValueError` when a non compatible data struct is fed into the constructor.

**Comment by amitprabhakar05** [comment]
Some languages have options in which they can fix the type of input,
especially in templates (same as in meta-programming contexts). I know that
in c++, when I searched, it's called "Type Safety". However, when I tried
searching for doing the same in python, it comes out that, it's not
designed to do such checks. However if you still want, here is the link of
one such discussion:

http://stackoverflow.com/questions/3749796/type-safety-in-python

On Sat, May 2, 2015 at 4:22 PM, Sauvik Biswas notifications@github.com
wrote:

> One way is to convert it and check if all entries are float. That way we
> can avoid linearisation. In fact, linearisation might not be necessary. The
> code should raise a ValueError when a non compatible data struct is fed
> into the constructor.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/sauvikbiswas/vectorlib/issues/5#issuecomment-98330262
> .

## 

# 

Amit Prabhakar

# Singapore.

**Comment by sauvikbiswas** [comment]
The constructor deliberately uses duck typing http://en.wikipedia.org/wiki/Duck_typing http://en.wikipedia.org/wiki/Duck_typing. This ensures that anything that has a float rep becomes an element in the vector. Python is designed to use this property to max. If you are learning LISP, you’ll soon find out how handy duck typing is. In fact, all functional approaches have this tendency. vectorlib is written in functional form (almost 90% of it).

I have a solution, I have written about it. Let me implement it and see.

EDIT: In the same link you had sent, an answer specifically touches upon the point of duck typing.

**Comment by sauvikbiswas** [comment]
I have updated the constructor. If the input is not an 1D iterator, float mapping is not possible (catches `TypeError`). If it is a 1D iterator of un-floatable data, it throws `ValueError` as usual. This seems to have solved the problem. I have added the testing of these errors in the `unittests.py` file.

Also, I have removed the destructor. It was throwing a warning. Then I realised that a native destructor is already inherited from base Python `object`, when I declared `class vector(object)`.
